### PR TITLE
add yolox model

### PR DESCRIPTION
**Pull request by Mark-ZhouWX**
Thank you for your contribution to the MindYOLO repo.
Before submitting this PR, please make sure:

- [x] You have read the [Contributing Guidelines on pull requests](https://github.com/mindspore-lab/mindyolo/blob/master/CONTRIBUTING.md)
- [x] Your code builds clean without any errors or warnings
- [x] You are using approved terminology
- [ ] You have added unit tests

## Motivation

1. add yolox model and its tiny scale
2. update config yaml log
3. fix some bugs in validation while train mode
4. fix some bugs in loading optimizer ckpt

## Test Plan

(How should this PR be tested? Do you require special setup to run the test or repro the fixed bug?)

## Related Issues and PRs

(Is this PR part of a group of changes? Link the other relevant PRs and Issues here. Use https://help.github.com/en/articles/closing-issues-using-keywords for help on GitHub syntax)

### CODE REVIEW

File: mindyolo/data/dataset.py [hunk 1]
```diff
@@ -243,6 +243,11 @@ def __getitem__(self, index):
                         self.albumentations = Albumentations(size=self.img_size)
                     image, labels = self.albumentations(image, labels, **_trans)
                 else:
+                    if image is None:
+                        image, hw_ori = self.load_image(index)
+                        labels = self.labels[index].copy()
+                        new_shape = self.img_size if not self.rect else self.batch_shapes[self.batch[index]]
+                        image, labels, hw_ori, hw_scale, pad = self.letterbox(image, labels, hw_ori, new_shape,)
                     image, labels = getattr(self, func_name)(image, labels, **_trans)
```
Comment (zhanghuiyao) [inline mindyolo/data/dataset.py:251]: 这个可以不用加吧 是否可以直接在配置文件中增加"letterbox"操作

File: mindyolo/models/__init__.py [hunk 2]
```diff
@@ -20,6 +20,7 @@
 from .losses import *
 from .yolov8 import *
 from .yolov7 import *
+from .yolox import *
```
Comment (zhanghuiyao) [inline mindyolo/models/__init__.py:23]: 这个位置可以放到最后，前面的数字版本连起来美观一些

File: mindyolo/models/heads/yolox_head.py [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+    @staticmethod
+    def _make_grid(nx=20, ny=20, dtype=ms.float32):
+        # meshgrid is not supported on a specific model of machine
+        # an alternative solution is adopted, which will be updated later
+        x = mnp.arange(nx).astype(ms.int32)
+        y = mnp.arange(ny).astype(ms.int32)
+        xv = ops.tile(x.reshape(1, nx), (ny, 1))
+        yv = ops.tile(y.reshape(ny, 1), (1, nx))
+        #xv, yv = ops.meshgrid((mnp.arange(nx), mnp.arange(ny)))
+        return ops.cast(ops.stack((xv, yv), 2).view((1, 1, ny, nx, 2)), dtype)
```
Comment (zhanghuiyao) [inline mindyolo/models/heads/yolox_head.py:113]: 文件末尾添加空行，全局检查一下

File: mindyolo/models/layers/bottleneck.py [hunk 4]
```diff
@@ -1,15 +1,15 @@
 from mindspore import nn, ops
 
-from .conv import ConvNormAct
+from .conv import ConvNormAct, DWConvNormAct
 
 
 class Bottleneck(nn.Cell):
     # Standard bottleneck
-    def __init__(self, c1, c2, shortcut=True, k=(1, 3), g=(1, 1), e=0.5, momentum=0.97, eps=1e-3, sync_bn=False):  # ch_in, ch_out, shortcut, kernels, groups, expand
+    def __init__(self, c1, c2, shortcut=True, k=(1, 3), g=(1, 1), e=0.5, act=True, momentum=0.97, eps=1e-3, sync_bn=False):  # ch_in, ch_out, shortcut, groups, kernels, expand
```
Comment (zhanghuiyao) [inline mindyolo/models/layers/bottleneck.py:8]: 注释部分 groups, kernels 顺序跟参数顺序对应

File: mindyolo/models/losses/yolox_loss.py [hunk 5]
```diff
@@ -0,0 +1,313 @@
...
+            anchor_strides [..., 1] /= self.input_size[1]
+
+        return anchor_center_pos, anchor_strides
+
+    def in_box(self, anchors, boxes):
+        # splitted_diff = ops.split(ops.concat([anchors - boxes[..., :2], boxes[..., 2:] - anchors],axis=-1),
+        #                           axis=-1, output_num=4)
+        # temp1 = ops.logical_and(splitted_diff[0] > 0.0, splitted_diff[1] > 0.0)
+        # temp2 = ops.logical_and(splitted_diff[2] > 0.0, splitted_diff[3] > 0.0)
+        # in_mask = ops.logical_and(temp1, temp2).squeeze(-1)
```
Comment (zhanghuiyao) [inline mindyolo/models/losses/yolox_loss.py:104]: 无效注释删除，全局检查一下

File: mindyolo/models/model_factory.py [hunk 6]
```diff
@@ -92,13 +92,16 @@ def parse_model(d, ch, nc, sync_bn=False):  # model_dict, input_channels(3)
     _SYNC_BN = sync_bn
     if _SYNC_BN:
         logger.info('Parse model with Sync BN.')
-    # logger.info('\n%3s%18s%3s%10s  %-40s%-30s' % ('', 'from', 'n', 'params', 'module', 'arguments'))  # print detail
+    logger.info('')
+    logger.info('network structure are as follows')
+    logger.info('%3s%18s%3s%10s  %-60s%-40s' % ('', 'from', 'n', 'params', 'module', 'arguments'))
```
Comment (zhanghuiyao) [inline mindyolo/models/model_factory.py:97]: 网络结构部分打印篇幅较长，对初级用户不友好，默认不打印，保留原注释形式

File: mindyolo/models/model_factory.py [hunk 7]
```diff
@@ -155,12 +160,16 @@ def parse_model(d, ch, nc, sync_bn=False):  # model_dict, input_channels(3)
 
         t = str(m) # module type
         np = sum([x.size for x in m_.get_parameters()])  # number params
+        np_trainable = sum([x.size for x in m_.trainable_params()])  # number trainable params
+        num_total_param += np
+        num_train_param += np_trainable
         m_.i, m_.f, m_.type, m_.np = i, f, t, np  # attach index, 'from' index, type, number params
         layers_param.append((i, f, t, np))
-        # logger.info('%3s%18s%3s%10.0f  %-40s%-30s' % (i, f, n, np, t, args))  # print detail
+        logger.info('%3s%18s%3s%10.0f  %-60s%-40s' % (i, f, n, np, t, args+[kwargs] if kwargs else args))  # print
```
Comment (zhanghuiyao) [inline mindyolo/models/model_factory.py:168]: 同上

File: mindyolo/models/utils/box_ops.py [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+        boxes (Tensor[N, 4]): boxes are specified by their (x1, y1, x2, y2) coordinates
+        clip_size (height, width): The clipping box's size.
+    """
+    h, w = clip_size
+    x1 = boxes[..., 0].clip(0, w)
+    y1 = boxes[..., 1].clip(0, h)
+    x2 = boxes[..., 2].clip(0, w)
+    y2 = boxes[..., 3].clip(0, h)
+    boxes = ops.stack((x1, y1, x2, y2), axis=-1)
+    return boxes
```
Comment (zhanghuiyao) [inline mindyolo/models/utils/box_ops.py:72]: 结尾空行

File: mindyolo/models/utils/box_ops.py [hunk 9]
```diff
@@ -0,0 +1,72 @@
+from typing import Tuple
```
Comment (zhanghuiyao) [inline mindyolo/models/utils/box_ops.py:1]: 这个文件的操作是否可以直接使用已有 utils 里面的操作替代

File: mindyolo/data/dataset.py (hunk 1 above)
Comment (Mark-ZhouWX) [reply mindyolo/data/dataset.py:251]: 需要加，配置文件中mosaic是以0.5概率出现的；

File: mindyolo/models/utils/box_ops.py (hunk 9 above)
Comment (Mark-ZhouWX) [reply mindyolo/models/utils/box_ops.py:1]: 这个文件里的操作是公共的，已有的utils里没有；但有些私有py里有，比如cxywh2xyxy，在yolov8_loss.py里有

File: mindyolo/models/__init__.py (hunk 2 above)
Comment (Mark-ZhouWX) [reply mindyolo/models/__init__.py]: 已修改

File: mindyolo/models/losses/yolox_loss.py (hunk 5 above)
Comment (Mark-ZhouWX) [reply mindyolo/models/losses/yolox_loss.py]: 已检查

File: mindyolo/models/model_factory.py (hunk 6 above)
Comment (Mark-ZhouWX) [reply mindyolo/models/model_factory.py]: 已修改

File: mindyolo/models/utils/box_ops.py (hunk 8 above)
Comment (Mark-ZhouWX) [reply mindyolo/models/utils/box_ops.py]: 已全局检查

File: mindyolo/models/model_factory.py (hunk 7 above)
Comment (Mark-ZhouWX) [reply mindyolo/models/model_factory.py]: 已修改

File: mindyolo/models/heads/yolox_head.py (hunk 3 above)
Comment (Mark-ZhouWX) [reply mindyolo/models/heads/yolox_head.py]: 已修改

File: mindyolo/models/layers/bottleneck.py (hunk 4 above)
Comment (Mark-ZhouWX) [reply mindyolo/models/layers/bottleneck.py]: 已修改

File: configs/yolox/hyp.scratch.yaml [hunk 10]
```diff
@@ -0,0 +1,45 @@
...
+  group_param: filter_bias_and_bn # group param strategy
+  gp_weight_decay: 0.0005  # group param weight decay 5e-4
+
+loss:
+   name: YOLOxLoss
+   input_size: [640, 640]
+
+#
+#ms_mode: 1
+#ms_jit: False
```
[the remark this answers is not in the document]
Comment (Mark-ZhouWX) [reply configs/yolox/hyp.scratch.yaml:15]: 已删除

File: mindyolo/models/utils/box_ops.py (hunk 9 above)
Comment (Mark-ZhouWX) [reply mindyolo/models/utils/box_ops.py:1]: 已调整

File: README.md [hunk 11]
```diff
@@ -39,7 +39,8 @@ Seed [MODEL ZOO](MODEL_ZOO.md).
 - [x] [YOLOv3](configs/yolov3)
 - [x] [YOLOv4](configs/yolov4)
 - [ ] [YOLOv6](configs/yolov6)
-- [ ] [YOLOX](configs/yolox)
+- [x] [YOLOX](configs/yolox)
+
```
[the remark this answers is not in the document]
Comment (Mark-ZhouWX) [reply README.md:43]: 已调整

File: configs/yolox/hyp.scratch.yaml [hunk 12]
```diff
@@ -0,0 +1,45 @@
...
+sync_bn: False
+ms_amp_level: O0
+keep_loss_fp32: True
+
+data:
+  num_parallel_workers: 4
+
+  train_transforms:
+    - {func_name: mosaic, prob: 1.0, mosaic9_prob: 0.0, translate: 0.1, scale: 0.9, degrees: 10.0, shear: 2.0}
+    - {func_name: mixup, prob: 1.0, alpha: 8.0, beta: 8.0, needed_mosaic: True } # TODO chage to yolox mixup
```
Comment (CaitinZhao) [inline configs/yolox/hyp.scratch.yaml:31]: Is there anything different from yolox mixup？

File: configs/yolox/hyp.scratch.yaml [hunk 13]
```diff
@@ -0,0 +1,45 @@
...
+    - {func_name: hsv_augment, prob: 1.0, hgain: 0.015, sgain: 0.7, vgain: 0.4} # TODO change to yolox hsv
```
Comment (CaitinZhao) [inline configs/yolox/hyp.scratch.yaml:32]: ditto

File: configs/yolox/hyp.scratch.yaml [hunk 14]
```diff
@@ -0,0 +1,45 @@
...
+    - {func_name: fliplr, prob: 0.5}
+    - {func_name: label_pad, padding_size: 160, padding_value: -1}
+    - {func_name: image_norm, scale: 255.}
+    - {func_name: image_transpose, bgr2rgb: True, hwc2chw: True}
+
+  test_transforms:
+    - {func_name: letterbox, scaleup: False}
+    - {func_name: label_norm, xyxy2xywh_: True}
+    - {func_name: label_pad, padding_size: 160, padding_value: -1}
+    - {func_name: image_norm, scale: 255. } # TODO megvii代码没有加这个norm
```
Comment (CaitinZhao) [inline configs/yolox/hyp.scratch.yaml:43]: 删除中文注释

File: configs/yolox/yolox-darknet53-2nd.yaml [hunk 15]
```diff
@@ -0,0 +1,63 @@
+__BASE__: [
+  '../coco.yaml',
+  './hyp.scratch.yaml',
+  './yolox-s-2nd.yaml'
+]
+
+weight: /home/zhouwuxing/projects/mindyolo/runs/2023.04.28-17:46:25/weights/yolox-darknet53-285_1848.ckpt
```
Comment (CaitinZhao) [inline configs/yolox/yolox-darknet53-2nd.yaml:7]: 不要用个人路径

File: configs/yolox/yolox-m-2nd.yaml [hunk 16]
```diff
@@ -0,0 +1,15 @@
+__BASE__: [
+  '../coco.yaml',
+  './hyp.scratch.yaml',
+  './yolox-s-2nd.yaml'
+]
+
+weight: /home/zhouwuxing/projects/mindyolo/runs/2023.04.24-09:35:19/weights/yolox-s-285_1848.ckpt
```
Comment (CaitinZhao) [inline configs/yolox/yolox-m-2nd.yaml:7]: ditto

File: configs/yolox/yolox-nano-2nd.yaml [hunk 17]
```diff
@@ -0,0 +1,63 @@
...
+  './hyp.scratch.yaml',
+  './yolox-s-2nd.yaml'
+]
+
+# difference with 2nd training of s
+# 1. img_size is 416
+# 2. depthwise is true
+# 3. depth and width
+
+weight: /home/zhouwuxing/projects/mindyolo/runs/2023.05.08-05:18:55/weights/yolox-nano-285_1848.ckpt
```
Comment (CaitinZhao) [inline configs/yolox/yolox-nano-2nd.yaml:12]: ditto

File: configs/yolox/yolox-s-2nd.yaml [hunk 18]
```diff
@@ -0,0 +1,76 @@
+__BASE__: [
+  '../coco.yaml',
+  './hyp.scratch.yaml',
+]
+
+weight: /home/zhouwuxing/projects/mindyolo/runs/2023.05.12-09:01:30/weights/yolox-s-285_1848.ckpt
```
Comment (CaitinZhao) [inline configs/yolox/yolox-s-2nd.yaml:6]: ditto

File: configs/yolox/yolox-tiny-2nd.yaml [hunk 19]
```diff
@@ -0,0 +1,25 @@
...
+  '../coco.yaml',
+  './hyp.scratch.yaml',
+  './yolox-s-2nd.yaml'
+]
+
+# difference with 2nd training of s
+# 1. img_size
+# 2. depth and width
+
+weight: /home/zhouwuxing/projects/mindyolo/runs/2023.05.17-09:29:38/weights/yolox-tiny-285_1848.ckpt
```
Comment (CaitinZhao) [inline configs/yolox/yolox-tiny-2nd.yaml:11]: ditto

File: configs/yolox/yolox-x-2nd.yaml [hunk 20]
```diff
@@ -0,0 +1,15 @@
+__BASE__: [
+  '../coco.yaml',
+  './hyp.scratch.yaml',
+  './yolox-s-2nd.yaml'
+]
+
+weight: /home/zhouwuxing/projects/mindyolo/runs/2023.04.24-09:35:19/weights/yolox-s-285_1848.ckpt
```
Comment (CaitinZhao) [inline configs/yolox/yolox-x-2nd.yaml:7]: ditto

File: mindyolo/models/losses/yolox_loss.py [hunk 21]
```diff
@@ -0,0 +1,305 @@
...
+        """
+        gt_valid_mask = targets[..., 1] >= 0  # defalut class column
+        gt_box_xyxy = box_cxcywh_to_xyxy(targets[:, :, 2:]) # (batch_size, gt_max, 4) in [xyxy] format
+        # reverse norm
+        gt_box_xyxy_raw = box_clip(box_scale(gt_box_xyxy, self.input_size), self.input_size)
+        # to cxcywh format
+        bbox_true = box_xyxy_to_cxcywh(gt_box_xyxy_raw)
+        is_inbox_and_incenter, pre_fg_mask = self._get_foreground(gt_box_xyxy_raw, gt_valid_mask)
+
+        batch_size = P.Shape()(preds)[0]
```
Comment (CaitinZhao) [inline mindyolo/models/losses/yolox_loss.py:165]: 统一刷一下，不要用P.大写字母开头的写法，全部改成Tensor.xxx或者ops.xxx这种

Comment (CaitinZhao) [inline mindyolo/models/losses/yolox_loss.py:165]: batch_size = preds.shape[0]

File: mindyolo/models/losses/yolox_loss.py [hunk 22]
```diff
@@ -0,0 +1,305 @@
...
+        # process label
+        gt_classes = F.cast(targets[:, :, 1:2].squeeze(-1), mindspore.int32)
+        pair_wise_ious = batch_box_iou(bbox_true, bbox_preds, xywh=True)  # (batch_size, gt_max, 8400)
+        pair_wise_ious = pair_wise_ious * pre_fg_mask
+        pair_wise_iou_loss = -P.Log()(pair_wise_ious + 1e-8) * pre_fg_mask
+        gt_classes_ = self.one_hot(gt_classes, self.num_class, self.on_value, self.off_value)
+        # (bs, num_gt_max, num_class) -> (bs, num_gt_max, num_total_anchor, num_class)
+        gt_classes_expaned = ops.repeat_elements(self.unsqueeze(gt_classes_, 2), rep=total_num_anchors, axis=2)
+        gt_classes_expaned = F.stop_gradient(gt_classes_expaned)
+        cls_preds_ = P.Sigmoid()(ops.repeat_elements(self.unsqueeze(cls_preds, 1), rep=gt_max, axis=1)) * \
```
Comment (CaitinZhao) [inline mindyolo/models/losses/yolox_loss.py:183]: Tensor.sigmoid()

File: mindyolo/models/losses/yolox_loss.py [hunk 23]
```diff
@@ -0,0 +1,305 @@
...
+        cls_target = P.BatchMatMul(transpose_a=True)(pos_mask, gt_classes_)
+
+        cls_target = cls_target * pred_ious_this_matching
+        obj_target = P.ReduceMax()(pos_mask, 1)  # (batch_size, 8400)
+
+        # calculate l1_target
+        reg_target = F.stop_gradient(reg_target)
+        cls_target = F.stop_gradient(cls_target)
+        obj_target = F.stop_gradient(obj_target)
+        bbox_preds = F.cast(bbox_preds, mindspore.float32)
```
Comment (CaitinZhao) [inline mindyolo/models/losses/yolox_loss.py:241]: bbox_preds = bbox_preds.astype(ms.float32)

File: mindyolo/models/losses/yolox_loss.py [hunk 24]
```diff
@@ -0,0 +1,305 @@
...
+            l1_preds = self.get_l1_format(bbox_preds)
+            l1_target = F.stop_gradient(l1_target)
+            l1_target = F.cast(l1_target, mindspore.float32)
+            l1_preds = F.cast(l1_preds, mindspore.float32)
+            loss_l1 = P.ReduceSum()(self.l1_loss(l1_preds, l1_target), -1) * obj_target
+            loss_l1 = P.ReduceSum()(loss_l1)
+        # calculate target -----------END-------------------------------------------------------------------------------
+        iou = bbox_iou(bbox_preds.reshape(-1, 4), reg_target.reshape(-1, 4), xywh=True).reshape(batch_size, -1)
+        loss_iou = (1 - iou * iou) * obj_target  # (bs, num_total_anchor)
+        loss_iou = P.ReduceSum()(loss_iou)
```
Comment (CaitinZhao) [inline mindyolo/models/losses/yolox_loss.py:259]: loss_iou = loss_iou.sum()

File: mindyolo/models/losses/yolox_loss.py [hunk 25]
```diff
@@ -0,0 +1,305 @@
...
+
+        return loss_all, ops.stop_gradient(ops.stack((loss_all, loss_iou, loss_obj, loss_cls, loss_l1)))
+
+    def get_l1_format_single(self, reg_target, stride, eps):
+        """ calculate L1 loss related """
+        reg_target = reg_target / stride
+        reg_target_xy = reg_target[:, :, :2]
+        reg_target_wh = reg_target[:, :, 2:]
+        reg_target_wh = P.Log()(reg_target_wh + eps)
+        return P.Concat(-1)((reg_target_xy, reg_target_wh))
```
Comment (CaitinZhao) [inline mindyolo/models/losses/yolox_loss.py:292]: ops.concat

File: mindyolo/utils/config.py [hunk 26]
```diff
@@ -115,6 +111,7 @@ def __init__(self, cfg_dict):
         super(Config, self).__init__()
         for k, v in cfg_dict.items():
             setattr(self, k, Config(v) if isinstance(v, dict) else v)
+        self.__dict__.update(cfg_dict)
```
[the remark this answers is not in the document]
Comment (Mark-ZhouWX) [reply mindyolo/utils/config.py:114]: 测试过，去掉不行。重写__getattr__方法能确保config[key]表达式能取到值，但不能将config类中所有属性进行注册。加上self.__dict__.update(cfg_dict)能将 嵌套的config属性注册进__str__

File: configs/yolox/hyp.scratch.yaml (hunk 12 above)
Comment (Mark-ZhouWX) [reply configs/yolox/hyp.scratch.yaml]: There are some slignt difference. The hsv_gain in the original version is a biad that adds to the img value, while it is multipying factor in mindyolo. Also, there is a random seletion button determins whether each of h, s, v channel are open or not.

File: configs/yolox/hyp.scratch.yaml (hunk 14 above)
Comment (Mark-ZhouWX) [reply configs/yolox/hyp.scratch.yaml]: resolved

File: configs/yolox/yolox-m-2nd.yaml (hunk 16 above)
Comment (Mark-ZhouWX) [reply configs/yolox/yolox-m-2nd.yaml]: resolved

File: configs/yolox/yolox-s-2nd.yaml (hunk 18 above)
Comment (Mark-ZhouWX) [reply configs/yolox/yolox-s-2nd.yaml]: resolved

File: mindyolo/models/losses/yolox_loss.py (hunk 21 above)
Comment (Mark-ZhouWX) [reply mindyolo/models/losses/yolox_loss.py]: resolved

File: mindyolo/models/losses/yolox_loss.py (hunk 22 above)
Comment (Mark-ZhouWX) [reply mindyolo/models/losses/yolox_loss.py]: resolved

File: mindyolo/models/losses/yolox_loss.py (hunk 23 above)
Comment (Mark-ZhouWX) [reply mindyolo/models/losses/yolox_loss.py]: resolved

File: mindyolo/models/losses/yolox_loss.py (hunk 25 above)
Comment (Mark-ZhouWX) [reply mindyolo/models/losses/yolox_loss.py]: resolved

File: mindyolo/models/losses/yolox_loss.py (hunk 24 above)
Comment (Mark-ZhouWX) [reply mindyolo/models/losses/yolox_loss.py]: resolved

File: mindyolo/models/heads/yolox_head.py [hunk 27]
```diff
@@ -0,0 +1,110 @@
...
+        return output_t  # bs, 6400, 85
+
+    @staticmethod
+    def _make_grid(nx=20, ny=20, dtype=ms.float32):
+        # meshgrid is not supported on a specific model of machine
+        # an alternative solution is adopted, which will be updated later
+        x = mnp.arange(nx).astype(ms.int32)
+        y = mnp.arange(ny).astype(ms.int32)
+        xv = ops.tile(x.reshape(1, nx), (ny, 1))
+        yv = ops.tile(y.reshape(ny, 1), (1, nx))
```
[the remark this answers is not in the document]
Comment (Mark-ZhouWX) [reply mindyolo/models/heads/yolox_head.py]: resolved

File: configs/yolox/yolox-nano-2nd.yaml (hunk 17 above)
Comment (Mark-ZhouWX) [reply configs/yolox/yolox-nano-2nd.yaml]: resolved

File: configs/yolox/yolox-tiny-2nd.yaml (hunk 19 above)
Comment (Mark-ZhouWX) [reply configs/yolox/yolox-tiny-2nd.yaml]: resolved

File: configs/yolox/yolox-x-2nd.yaml (hunk 20 above)
Comment (Mark-ZhouWX) [reply configs/yolox/yolox-x-2nd.yaml]: resolved

File: mindyolo/models/yolox.py [hunk 28]
```diff
@@ -0,0 +1,62 @@
+import math
+
+import mindspore as ms
+import numpy as np
+from mindspore import nn, Tensor
+
+from .initializer import initialize_defult
+
+__all__ = [
+    'YOLOx',
```
Comment (zhanghuiyao) [inline mindyolo/models/yolox.py:10]: YOLOx -> YOLOX

File: mindyolo/models/losses/yolox_loss.py [hunk 29]
```diff
@@ -0,0 +1,306 @@
...
+from mindyolo.models.registry import register_model
+from mindyolo.models.layers.utils import box_cxcywh_to_xyxy, box_xyxy_to_cxcywh, box_scale, box_clip
+
+__all__ = [
+    'YOLOxLoss',
+]
+
+
+@register_model
+class YOLOxLoss(nn.Cell):
```
Comment (zhanghuiyao) [inline mindyolo/models/losses/yolox_loss.py:17]: YOLOx -> YOLOX

File: mindyolo/models/heads/yolox_head.py [hunk 30]
```diff
@@ -0,0 +1,107 @@
...
+import mindspore as ms
+from mindspore import nn, ops, Tensor
+from mindspore import numpy as mnp
+from mindspore.common import initializer as init
+
+from mindyolo.models.layers.conv import ConvNormAct, DWConvNormAct
+from mindyolo.models.layers.utils import meshgrid
+
+
+class YOLOxHead(nn.Cell):
```
Comment (zhanghuiyao) [inline mindyolo/models/heads/yolox_head.py:12]: YOLOxHead -> YOLOXHead

File: mindyolo/models/losses/yolox_loss.py (hunk 29 above)
Comment (Mark-ZhouWX) [reply mindyolo/models/losses/yolox_loss.py]: resolved

File: mindyolo/models/heads/yolox_head.py (hunk 30 above)
Comment (Mark-ZhouWX) [reply mindyolo/models/heads/yolox_head.py]: resolved

**Review by zhanghuiyao — APPROVED** [review]

**Review by CaitinZhao — APPROVED** [review]

**Pull request merged**